third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -3236,6 +3243,156 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
         }
     }
 
+    fn call_to_borrow_global(
+        &mut self,
+        loc: &Loc,
+        e: &EA::Exp,
+        i: &EA::Exp,
+        mutable: bool,
+        expected_type: &Type,
+        context: &ErrorMessageContext,
+    ) -> ExpData {
+        fn convert_name_to_type(
```

Comment:
I'm not sure whether we need all this complexity here of creating builtin function names and sending them through the regular translate_call. Why not instead do the following:

```
let resource_opt = self.translate_exp_to_type(e);
if let Some(resource) = resource_opt { 
   // need to add constraint to `resource` that it has key, there should be examples for this
   <<check that resource is actually a resource>>
   let index = self.translate_exp(i, &Type::new_prim(PrimitiveType::Address);
   let node_id = ...; // Need to create new node_id with location and result type
   ExpData::Call(node_id, Operation::BorrowGlobal, vec![index]);
} else {
   <<report error>> // (currently you seem to just ignore of `e` is not a resource)
}
```

Similar for vector index. This can also be further refined to give better error messages, as pointed out above.

Reply:
I was trying to use the method you mentioned but was blocked by the logic on type instantiation on generic types. The regular `translate_call` instead helps unblock it quickly by handling all details there. To speed things up, I am wondering whether we can use this way in this PR and refactor it later?

Reply:
Instantiations are not difficult to deal with, though it gets a bit more difficult for vector indexing because this one has not Operation::VecBorrow but one must generate an Operation::Function and lookup the FunctionId in the builtin module. In general, we are not in a need of a quick hack.

Reply:
Some more info what you need to do regards instantiations. Every Call instruction contains two annotations: (a) the result type of the call (b) a vector with the instantiation of the functions.

1. For `Call(id, Operation::BorrowGlobal), arg)` the instantiations is simply `vec![resource_type]`. This should be straightforward.
2. For `Call(id, Operation::Function(vec_borrow), v, i)` the instantiation is `vec![element_type])`. You should check you are looking at a vector type here anyway for dealing with the error message problem I have flagged. You can check by specializing the type of `v` (`subs.specialize()`). If the type isn't known you produce an error `type of index operand need to be known at this time`. Rust does that too. I think in order to implement good error messages and also to do `v[i].x` (without `&`), you need to know the type in any case. 

All of what I'm saying can also be wrong ;-) Just guessing.

Reply:
And another remark in this context. In order to make `vec[x].f` work, you should probably not longer have a special rule for translating `&_[_]` but you translate the index operator directly. You'd end up translating `v[i]` into `*borrow_vector(v,i)`. If the user now adds a reference, you have `&*borrow_vector(v, i)`. The `&*` is removed by the bytecode generator [here](https://github.com/aptos-labs/aptos-core/blob/ebcdb747bad523db41248c6212e33f305e35f5d8/third_party/move/move-compiler-v2/src/bytecode_generator.rs#L1079-L1083)! So you get what you expect in the final bytecode.